include/knowhere/index/index_static.h

```diff
@@ -0,0 +1,107 @@
+     private:                                                                           \
+        template <typename T>                                                           \
+        static auto                                                                     \
+        test(int) -> decltype(T::func_name(std::declval<Args>()...), std::true_type{}); \
+                                                                                        \
+        template <typename>                                                             \
+        static auto                                                                     \
+        test(...) -> std::false_type;                                                   \
+                                                                                        \
+     public:                                                                            \
+        static constexpr bool value = decltype(test<C>(0))::value;                      \
+    };
+
+DEFINE_HAS_STATIC_FUNC(StaticCreateConfig)
+DEFINE_HAS_STATIC_FUNC(StaticEstimateLoadResource)
+DEFINE_HAS_STATIC_FUNC(StaticHasRawData)
+
+template <typename DataType>
+class IndexStaticFaced {
+ public:
+    static std::unique_ptr<BaseConfig>
+    CreateConfig(const knowhere::IndexType& indexType, const knowhere::IndexVersion& version);
+
+    static expected<Resource>
+    EstimateLoadResource(const knowhere::IndexType& indexType, const knowhere::IndexVersion& version,
+                         const float file_size, const knowhere::Json& params);
+
+    static bool
+    HasRawData(const knowhere::IndexType& indexType, const knowhere::IndexVersion& version,
+               const knowhere::Json& params);
+
+    template <typename VecIndexNode>
+    IndexStaticFaced&
+    RegisterStaticFunc(const knowhere::IndexType& indexType) {
+        static_assert(
+            has_static_StaticCreateConfig<VecIndexNode,
+                                          decltype(IndexStaticFaced<DataType>::InternalStaticCreateConfig)>::value,
+            "VecIndexNode must implement StaticCreateConfig function");
+        staticCreateConfigMap[indexType] = VecIndexNode::StaticCreateConfig;
+
+        if constexpr (has_static_StaticEstimateLoadResource<
+                          VecIndexNode, decltype(IndexStaticFaced<DataType>::InternalEstimateLoadResource)>::value) {
+            staticEstimateLoadResourceMap[indexType] = VecIndexNode::StaticEstimateLoadResource;
+        }
+
+        if constexpr (has_static_StaticHasRawData<
+                          VecIndexNode, decltype(IndexStaticFaced<DataType>::InternalStaticHasRawData)>::value) {
+            staticHasRawDataMap[indexType] = VecIndexNode::StaticHasRawData;
+        }
+
+        return Instance();
+    }
+
+    static IndexStaticFaced&
+    Instance();
+
+ private:
+    static expected<Resource>
+    InternalEstimateLoadResource(const float file_size, const knowhere::BaseConfig& config);
```

Comment:
`file_size` is `float`? O_o
why `file_size`, not just something like `uint64_t size_in_bytes` ?

Reply:
The unit is GB, which is common used between the milvus and knowhere to describe the data size